x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/highlight/SemanticTextHighlighter.java

```diff
@@ -0,0 +1,208 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.highlight;
+
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.KnnByteVectorQuery;
+import org.apache.lucene.search.KnnFloatVectorQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.elasticsearch.common.text.Text;
+import org.elasticsearch.common.xcontent.support.XContentMapValues;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper.DenseVectorFieldType;
+import org.elasticsearch.index.mapper.vectors.SparseVectorFieldMapper.SparseVectorFieldType;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.search.fetch.subphase.highlight.FieldHighlightContext;
+import org.elasticsearch.search.fetch.subphase.highlight.HighlightField;
+import org.elasticsearch.search.fetch.subphase.highlight.Highlighter;
+import org.elasticsearch.search.vectors.VectorData;
+import org.elasticsearch.xpack.core.ml.search.SparseVectorQueryWrapper;
+import org.elasticsearch.xpack.inference.mapper.SemanticTextField;
+import org.elasticsearch.xpack.inference.mapper.SemanticTextFieldMapper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link Highlighter} designed for the {@link SemanticTextFieldMapper}.
+ * This highlighter extracts semantic queries and evaluates them against each chunk produced by the semantic text field.
+ * It returns the top-scoring chunks as snippets, optionally sorted by their scores.
+ */
+public class SemanticTextHighlighter implements Highlighter {
+    public static final String NAME = "semantic";
+
+    private record OffsetAndScore(int offset, float score) {}
+
+    @Override
+    public boolean canHighlight(MappedFieldType fieldType) {
+        if (fieldType instanceof SemanticTextFieldMapper.SemanticTextFieldType) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public HighlightField highlight(FieldHighlightContext fieldContext) throws IOException {
+        SemanticTextFieldMapper.SemanticTextFieldType fieldType = (SemanticTextFieldMapper.SemanticTextFieldType) fieldContext.fieldType;
+        if (fieldType.getEmbeddingsField() == null) {
+            // nothing indexed yet
+            return null;
```

Comment:
Question: what is rendered in the response when we return `null` here? Is the `highlight.<field_name>` section of the response still rendered? If not, it may be confusing for users as the behavior is the same as if they did not request highlighting at all. Should we return a value that ensures that `highlight.<field_name>` is always rendered in the response, even the value is an empty array?

Reply:
It's on a per document basis so I think it makes sense to return null when there's no content for the semantic text field in the returned document. That's also consistent with the default highlighting behaviour.

Reply:
Ok, thanks for clarifying that this is consistent with the default highlighting behavior!